pkg/kvstore/allocator/allocator.go

```diff
@@ -396,6 +396,48 @@ func (k *kvstoreBackend) Release(ctx context.Context, key allocator.AllocatorKey
 	return nil
 }
 
+// RunLocksGC scans the kvstore for unused locks and removes them
+func (k *kvstoreBackend) RunLocksGC(ctx context.Context, staleKeysPrevRound map[string]kvstore.Value) (map[string]kvstore.Value, error) {
+	// fetch list of all /../locks keys
+	allocated, err := kvstore.ListPrefix(ctx, k.lockPrefix)
+	if err != nil {
+		return nil, fmt.Errorf("list failed: %s", err)
+	}
+
+	staleKeys := map[string]kvstore.Value{}
+
+	// iterate over /../locks
+	for key, v := range allocated {
+		scopedLog := log.WithFields(logrus.Fields{
+			fieldKey:     key,
+			fieldLeaseID: fmt.Sprintf("%x", v.LeaseID),
+		})
+		// Only delete if this key was previously marked as to be deleted
+		if modRev, ok := staleKeysPrevRound[key]; ok &&
+			modRev.ModRevision == v.ModRevision &&
+			modRev.LeaseID == v.LeaseID &&
+			modRev.SessionID == v.SessionID {
+			if err := kvstore.Delete(ctx, key); err != nil {
+				scopedLog.WithError(err).
+					Warning("Unable to unlocked distributed lock due client staleness." +
+						" Please check the connectivity between the KVStore and the client with that lease ID.")
```

Comment:
How does the user determine the client for that lease ID? (Can we document this somewhere and add a cilium.link for it?)

Reply:
Unfortunately it's tricky... One needs to access each cilium status on each cilium instance to know which lease ID the cilium-agent is using. I thought about this while developing the PR and the easiest would be to add it into the cilium node structure that we have. What do you think? Also, be aware this needs to be backported to 1.5 and 1.6

Reply:
I'm happy with "do this annoying thing by executing commands in a bunch of pods" for a safer backport.

It'd make a big difference for those users that hit such an issue to pinpoint where and tell them explicitly, but I'm less sure how invasive such a change is so probably worth tackling separately so we can review it closely.